OpenRA.Mods.Common/UpdateRules/Rules/20171014/DecoupleSelfReloading.cs

```diff
@@ -0,0 +1,84 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+
+namespace OpenRA.Mods.Common.UpdateRules.Rules
+{
+	public class DecoupleSelfReloading : UpdateRule
+	{
+		public override string Name { get { return "Replace 'SelfReloads' with 'ReloadAmmoPool'"; } }
+		public override string Description
+		{
+			get
+			{
+				return "'SelfReloads', 'SelfReloadDelay', 'ReloadCount', 'ResetOnFire'\n" +
+					"and 'RearmSound' were renamed and moved from 'AmmoPool' to a new 'ReloadAmmoPool' trait.";
+			}
+		}
+
+		public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
+		{
+			var poolNumber = 0;
+			var ammoPools = actorNode.ChildrenMatching("AmmoPool");
+			foreach (var pool in ammoPools)
+			{
+				var selfReloads = pool.LastChildMatching("SelfReloads");
+				if (selfReloads == null || !selfReloads.NodeValue<bool>())
+					continue;
+
+				poolNumber++;
+				var reloadOnCond = new MiniYamlNode("ReloadAmmoPool@" + poolNumber, "");
+
+				var name = pool.LastChildMatching("Name");
+				if (name != null)
+					reloadOnCond.AddNode("AmmoPool", name.NodeValue<string>());
+
+				var selfReloadDelay = pool.LastChildMatching("SelfReloadDelay");
+				if (selfReloadDelay != null)
+				{
+					selfReloadDelay.RenameKey("Delay");
+					reloadOnCond.AddNode(selfReloadDelay);
+					pool.RemoveNodes("SelfReloadDelay");
+				}
+
+				pool.RemoveNodes("SelfReloads");
+
+				var reloadCount = pool.LastChildMatching("ReloadCount");
+				if (reloadCount != null)
+				{
+					reloadCount.RenameKey("Count");
+					reloadOnCond.AddNode(reloadCount);
+					pool.RemoveNodes("ReloadCount");
+				}
+
+				var reset = pool.LastChildMatching("ResetOnFire");
+				if (reset != null)
+				{
+					reloadOnCond.AddNode(reset);
+					pool.RemoveNodes("ResetOnFire");
+				}
+
+				var rearmSound = pool.LastChildMatching("RearmSound");
+				if (rearmSound != null)
+				{
+					rearmSound.RenameKey("Sound");
+					reloadOnCond.AddNode(rearmSound);
+					pool.RemoveNodes("RearmSound");
+				}
+
+				actorNode.AddNode(reloadOnCond);
```

Comment:
```
DecoupleSelfReloading: Replace 'SelfReloads' with 'ReloadAmmoPool'
   Updating mod... FAILED

   The automated changes for this rule were not applied because of an error.
   After the issue reported below is resolved you should run the updater
   with SOURCE set to DecoupleSelfReloading to retry these changes

   The exception reported was:
     System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
       at System.ThrowHelper.ThrowInvalidOperationException (System.ExceptionResource resource) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2017-12/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/throwhelper.cs:104 
       at System.Collections.Generic.List`1+Enumerator[T].MoveNextRare () [0x00013] in <bb7b695b8c6246b3ac1646577aea7650>:0 
       at System.Collections.Generic.List`1+Enumerator[T].MoveNext () [0x0004a] in <bb7b695b8c6246b3ac1646577aea7650>:0 
       at System.Linq.Enumerable+WhereListIterator`1[TSource].MoveNext () [0x0004e] in /Users/builder/jenkins/workspace/build-package-osx-mono/2017-12/external/bockbuild/builds/mono-x64/external/corefx/src/System.Linq/src/System/Linq/Where.cs:348 
       at OpenRA.Mods.Common.UpdateRules.Rules.DecoupleSelfReloading+<UpdateActorNode>c__Iterator0.MoveNext () [0x001ab] in /Users/paul/src/OpenRA/OpenRA.Mods.Common//UpdateRules/Rules/20171014/DecoupleSelfReloading.cs:32 
       at OpenRA.Mods.Common.UpdateRules.UpdateUtils+<ApplyTopLevelTransform>c__Iterator3.MoveNext () [0x00127] in /Users/paul/src/OpenRA/OpenRA.Mods.Common//UpdateRules/UpdateUtils.cs:220 
       at System.Collections.Generic.List`1[T].InsertRange (System.Int32 index, System.Collections.Generic.IEnumerable`1[T] collection) [0x000ea] in <bb7b695b8c6246b3ac1646577aea7650>:0 
       at System.Collections.Generic.List`1[T].AddRange (System.Collections.Generic.IEnumerable`1[T] collection) [0x00000] in <bb7b695b8c6246b3ac1646577aea7650>:0 
       at OpenRA.Mods.Common.UpdateRules.UpdateUtils.UpdateMod (OpenRA.ModData modData, OpenRA.Mods.Common.UpdateRules.UpdateRule rule, System.Collections.Generic.List`1[System.Tuple`3[OpenRA.FileSystem.IReadWritePackage,System.String,System.Collections.Generic.List`1[OpenRA.MiniYamlNode]]]& files, System.Collections.Generic.HashSet`1[T] externalFilenames) [0x00265] in /Users/paul/src/OpenRA/OpenRA.Mods.Common//UpdateRules/UpdateUtils.cs:173 
       at OpenRA.Mods.Common.UtilityCommands.UpdateModCommand.ApplyRules (OpenRA.ModData modData, System.Collections.Generic.IEnumerable`1[T] rules, System.Boolean skipMaps) [0x00081] in /Users/paul/src/OpenRA/OpenRA.Mods.Common//UtilityCommands/UpdateModCommand.cs:152 
```

This either needs to move outside the loop, or `ammoPools` needs to be `ToList()`ed so that you aren't enumerating the same collection you're modifying.

Reply:
Fixed by going the `ToList()` route.